src/Controller/CRUDController.php

```diff
@@ -1359,15 +1361,51 @@ protected function handleXmlHttpRequestErrorResponse(Request $request, FormInter
             return $this->renderJson([], Response::HTTP_NOT_ACCEPTABLE);
         }
 
-        $errors = [];
-        foreach ($form->getErrors(true) as $error) {
-            $errors[] = $error->getMessage();
+        return $this->json(
+            $this->buildConstraintViolationList($form),
+            Response::HTTP_BAD_REQUEST
+        );
+    }
+
+    private function buildConstraintViolationList(FormInterface $form): ConstraintViolationList
```

Comment:
How should I split this code outside CRUD?

It will be a php class without dependencies AFAIK.

It can be:

1. `Sonata\AdminBundle\Form\ConstraintViolationBuilder`?
2. `Sonata\AdminBundle\Form\ConstraintViolationListBuilder`?
3. `Sonata\AdminBundle\Form\FormErrorsToConstraintViolationList`?

would you make it static, or a full service injected? I guess for testability it would be better to be a service, but dunno.

Reply:
I dunno what's the best.

Could been something like https://github.com/sonata-project/SonataAdminBundle/blob/4.x/src/Util/TraversableToCollection.php

So `FormErrorsToConstraintViolationList` ; and I think a static is ok.
It will avoid to add a dependency...

Reply:
Changed to `FormErrorIteratorToConstraintViolationList`